lib/integrations/externalcloudaudit/error_counter.go

```diff
@@ -0,0 +1,403 @@
+	UpsertClusterAlert(ctx context.Context, alert types.ClusterAlert) error
+	// DeleteClusterAlert deletes the cluster alert with the specified ID.
+	DeleteClusterAlert(ctx context.Context, alertID string) error
+}
+
+// ErrorCounter is used when the External Audit Storage feature is enabled to
+// store audit events and session recordings on external infrastructure. It
+// effectively provides audit middlewares that count errors and raise or clear
+// cluster alerts based on recent error rates. Cluster alerts are used to
+// surface this information because Cloud customers don't have access to their
+// own Auth server logs.
+type ErrorCounter struct {
+	emits     errorCount
+	searches  errorCount
+	uploads   errorCount
+	downloads errorCount
+
+	alertService ClusterAlertService
+	clock        clockwork.Clock
+
+	categories []errorCategory
+}
+
+// NewErrorCounter takes a ClusterAlertService that will be used to raise or
+// clear cluster alerts and returns a new ErrorCounter.
+func NewErrorCounter(alertService ClusterAlertService) *ErrorCounter {
+	c := &ErrorCounter{
+		alertService: alertService,
+		clock:        clockwork.NewRealClock(),
+	}
+	c.categories = []errorCategory{
+		{
+			alertName:    eventEmitFailureClusterAlert,
+			alertMessage: eventEmitFailureClusterAlertMsgTemplate,
+			// Raise a cluster alert if the recent error rate reaches 40%.
+			alertThreshold: 0.4,
+			// Don't alert until there have been at least 10 errors.
+			minErrorsForAlert: 10,
+			// Clear the alert if the recent error rate is below 5%.
+			clearAlertThreshold: 0.05,
+			// Don't clear the alert unless there have been at least 10
+			// successes.
+			minSuccessesForClear: 10,
+			count:                &c.emits,
+		},
+		{
+			alertName:            eventSearchFailureClusterAlert,
+			alertMessage:         eventSearchFailureClusterAlertMsgTemplate,
+			minErrorsForAlert:    4,
+			alertThreshold:       0.4,
+			minSuccessesForClear: 4,
+			clearAlertThreshold:  0.05,
+			count:                &c.searches,
+		},
+		{
+			alertName:            sessionUploadFailureClusterAlert,
+			alertMessage:         sessionUploadFailureClusterAlertMsgTemplate,
+			minErrorsForAlert:    4,
+			alertThreshold:       0.4,
+			minSuccessesForClear: 8,
+			clearAlertThreshold:  0.05,
+			count:                &c.uploads,
+		},
+		{
+			alertName:            sessionDownloadFailureClusterAlert,
+			alertMessage:         sessionDownloadFailureClusterAlertMsgTemplate,
+			minErrorsForAlert:    4,
+			alertThreshold:       0.4,
+			minSuccessesForClear: 2,
+			clearAlertThreshold:  0.05,
+			count:                &c.downloads,
+		},
+	}
+	return c
+}
+
+// WrapAuditLogger returns an [events.AuditLogger] that will forward all calls
+// to [wrapped] and observe all errors encountered.
+func (c *ErrorCounter) WrapAuditLogger(wrapped events.AuditLogger) events.AuditLogger {
+	return newErrorCountingLogger(wrapped, &c.emits, &c.searches)
+}
```

Comment:
My reasoning at the time was that I'd prefer not to export the concrete types, because no other package should care about them, and the IsErrorCounter hack exists because TestAthenaAuditLogSetup is a hacky test, and I'd rather not start exporting types just to satisfy a hacky test in another package

But looking back that reasoning is probably flawed, I should just return concrete exported types, I'll go ahead and make that change